pkg/crosscluster/logical/batch_handler_test.go

```diff
@@ -0,0 +1,391 @@
+	for i, col := range tableDesc.PublicColumns() {
+		colMap.Set(col.GetID(), i)
+	}
+	return &EventBuilder{
+		t:         t,
+		tableDesc: tableDesc,
+		colMap:    colMap,
+		codec:     keys.SystemSQLCodec,
+	}
+}
+
+func (b *EventBuilder) encodeRow(timestamp hlc.Timestamp, row tree.Datums) roachpb.KeyValue {
+	indexEntries, err := rowenc.EncodePrimaryIndex(
+		b.codec,
+		b.tableDesc,
+		b.tableDesc.GetPrimaryIndex(),
+		b.colMap,
+		row,
+		false,
+	)
+	require.NoError(b.t, err)
+	require.Len(b.t, indexEntries, 1)
+	kv := roachpb.KeyValue{
+		Key:   indexEntries[0].Key,
+		Value: indexEntries[0].Value,
+	}
+	kv.Value.Timestamp = timestamp
+	kv.Value.InitChecksum(kv.Key)
+	return kv
+}
+
+// insertEvent creates an insert event for the given row at the specified timestamp.
+func (b *EventBuilder) insertEvent(time hlc.Timestamp, row tree.Datums) streampb.StreamEvent_KV {
+	event := streampb.StreamEvent_KV{
+		KeyValue: b.encodeRow(time, row),
+	}
+	return event
+}
+
+// updateEvent creates an update event for the given row and previous values at the specified timestamp.
+func (b *EventBuilder) updateEvent(
+	time hlc.Timestamp, row tree.Datums, prevValue tree.Datums,
+) streampb.StreamEvent_KV {
+	kv := b.encodeRow(time, row)
+	kvPrev := b.encodeRow(time, prevValue)
+	return streampb.StreamEvent_KV{
+		KeyValue:  kv,
+		PrevValue: kvPrev.Value,
+	}
+}
+
+// deleteEvent creates a delete event for the given row at the specified timestamp.
+func (b *EventBuilder) deleteEvent(
+	time hlc.Timestamp, prevValue tree.Datums,
+) streampb.StreamEvent_KV {
+	kv := b.encodeRow(time, prevValue)
+	return streampb.StreamEvent_KV{
+		KeyValue: roachpb.KeyValue{
+			Key:   kv.Key,
+			Value: roachpb.Value{Timestamp: time},
+		},
+		PrevValue: kv.Value,
+	}
+}
+
+// newBatchHandler creates a new batch handler for testing.
+func newBatchHandler(
+	t *testing.T, s serverutils.ApplicationLayerInterface, tableName string,
+) (*kvRowProcessor, catalog.TableDescriptor) {
+	ctx := context.Background()
+	desc := cdctest.GetHydratedTableDescriptor(t, s.ExecutorConfig(), tree.Name(tableName))
+	sd := sql.NewInternalSessionData(ctx, s.ClusterSettings(), "" /* opName */)
+	handler, err := newKVRowProcessor(
+		ctx,
+		&execinfra.ServerConfig{
+			DB:           s.InternalDB().(descs.DB),
+			LeaseManager: s.LeaseManager(),
+			Settings:     s.ClusterSettings(),
+		},
+		&eval.Context{
+			Codec:            s.Codec(),
+			Settings:         s.ClusterSettings(),
+			SessionDataStack: sessiondata.NewStack(sd),
+		},
+		execinfrapb.LogicalReplicationWriterSpec{},
+		map[descpb.ID]sqlProcessorTableConfig{
+			desc.GetID(): {
+				srcDesc: desc,
+			},
+		},
+	)
+	require.NoError(t, err)
+	return handler, desc
+}
+
+// addAndRemoveColumn is a metamorphic test option that randomly decides whether
+// to add and remove a column to trigger a rebuild of the primary key.
+var addAndRemoveColumn = metamorphic.ConstantWithTestBool("batch-handler-exhaustive-rebuild-primary-index", false)
```

Comment:
when we run a test under stress, i think these values will be the same for each stress run since these values get set at build time. If you instead init these values with an rng in the test, i think a stress run would cover all values, so long as a different seed is chosen on each test run.